src/msw/renderer.cpp

```diff
@@ -1093,12 +1099,86 @@ void wxRendererXP::DrawItemText(wxWindow* win,
         else
             textFlags |= DT_LEFT;
 
-        if ( align & wxALIGN_BOTTOM )
-            textFlags |= DT_BOTTOM;
-        else if ( align & wxALIGN_CENTER_VERTICAL )
-            textFlags |= DT_VCENTER;
+        /*
+        Bottom (DT_BOTTOM) and centered vertical (DT_VCENTER) alignment
+        are documented to be only used with the DT_SINGLELINE value which
+        doesn't handle multi-lines. In case of drawing multi-lines with
+        such alignment use DT_TOP (0), which does work for multi-lines,
+        and deal with the actual desired alignment ourselves with the help
+        of GetThemeTextExtent.
+        */
+        const bool requiresTopDrawing =
+            s_GetThemeTextExtent
+            && ( align & (wxALIGN_BOTTOM | wxALIGN_CENTRE_VERTICAL) ) != 0
+            && text.Contains(wxS('\n'));
+
+        if ( !requiresTopDrawing )
+        {
+            if ( align & wxALIGN_BOTTOM )
+                textFlags |= DT_BOTTOM | DT_SINGLELINE;
+            else if ( align & wxALIGN_CENTER_VERTICAL )
+                textFlags |= DT_VCENTER | DT_SINGLELINE;
```

Comment:
Because Windows XP doesn't have `GetThemeTextExtent()` this code gets used and while the addition of `DT_SINGLELINE` results in bottom and centered vertical aligned text to work it also renders multi-lines as a single line now. Formerly all text was always top aligned vertically whcih does render multi-lines correctly. I think here it should either:

a) (preferred) Use `DT_TOP` (without `DT_SINGLELINE`) if non-top vertical alignment is used and text contains a newline. I don't like the cost of checking the latter (which isn't done any longer with e19f903) but at least multi-lines then work as before: always top aligned but otherwise rendered correctly and of course single-line non-top aligned text gets properly aligned now because of using `DT_SINGLELINE`.

or:

b) Don't use `DT_SINGLELINE` at all like before, restoring previous behaviour under XP only: all text is top aligned.

Reply:
I'm fine with either (a) or (b). I'd like to say we don't need to care about XP at all any longer, but there are probably still people using it...

Reply:
Sorry this can't actually apply to XP as it doesn't have `DrawThemeTextEx()`, so it's only regarding what to do if `GetThemeTextExtent()` fails which I haven't seen happening yet so I'll keep things as-is right now and maybe change it after the code settles down more.